File: rq1/scripts/table_13.R
library(CausalImpact)
library(zoo)

DATA_DIR <- "rq1/data"
MILESTONES <- list(Oct2023 = as.Date("2023-10-23"), Mar2025 = as.Date("2025-03-17"), Jul2025 = as.Date("2025-07-21"))

snap_to_week <- function(target, weeks) {
  candidates <- weeks[weeks >= target]
  if (length(candidates) > 0) return(min(candidates))
  return(max(weeks[weeks < target]))
}

run_ci <- function(y, weeks, treatment_target) {
  td <- snap_to_week(treatment_target, weeks)
  pre_end <- td - 7; post_end <- td + 7 * 7
  idx <- weeks >= min(weeks) & weeks <= post_end
  s <- zoo(y[idx], order.by = weeks[idx])
  if (length(unique(y[idx])) <= 1) return(c(NA, NA))
  set.seed(20260827)
  tryCatch({
    ci <- CausalImpact(s, c(min(weeks), pre_end), c(td, post_end))
    avg <- ci$summary["Average", ]
    c(round(avg[["RelEffect"]] * 100, 1), round(avg[["p"]], 4))
  }, error = function(e) c(NA, NA))
}

weekly <- read.csv(file.path(DATA_DIR, "table13_weekly_series.csv"), stringsAsFactors = FALSE)
weekly$week <- as.Date(weekly$week)
groups <- c("VPN", "Politics")
filters <- c("Unmatched (non-UK)", "Unmatched", "Unmatched (UK)", "Raw", "Raw (UK)",
             "Matched", "Matched (UK)", "Classified", "Classified (UK)")
metrics <- c("posts", "comments", "total")

total <- length(groups) * length(filters) * length(metrics) * length(MILESTONES)

results <- data.frame(); count <- 0
for (grp in groups) {
  for (filt in filters) {
    sub <- weekly[weekly$group == grp & weekly$filter == filt, ]
    sub <- sub[order(sub$week), ]
    if (nrow(sub) == 0) next
    for (ms in names(MILESTONES)) {
      for (met in metrics) {
        count <- count + 1
        cat(sprintf("[%3d/%d] %s | %s | %s | %s ... ", count, total, grp, filt, met, ms))
        r <- run_ci(sub[[met]], sub$week, MILESTONES[[ms]])
        if (!is.na(r[1])) {
          sig <- ifelse(r[2] < 0.05, "**", ifelse(r[2] < 0.10, "*", ""))
          cat(sprintf("%+.0f%%%s\n", r[1], sig))
        } else cat("NA\n")
        results <- rbind(results, data.frame(group = grp, filter = filt, metric = met,
          milestone = ms, relative_effect_pct = r[1], p_value = r[2]))
      }
    }
  }
}
write.csv(results, file.path(DATA_DIR, "table13_reproduced.csv"), row.names = FALSE)
cat(sprintf("\nSaved: table13_reproduced.csv (%d rows)\n", nrow(results)))

File: rq1/scripts/table_13.py
"""
reproduce_table13.py — CausalImpact on all Table 13 filter levels.

162 models: 2 groups × 9 filters × 3 milestones × 3 metrics.

Input:  rq1/data/table13_weekly_series.csv
Output: rq1/data/table13_reproduced.csv

Usage:
  python rq1/reproduce_table13.py
"""
import os, warnings
os.environ["TF_CPP_MIN_LOG_LEVEL"] = "3"
warnings.filterwarnings("ignore")
import pandas as pd

if not hasattr(pd.DataFrame, "applymap"):
    pd.DataFrame.applymap = pd.DataFrame.map
from causalimpact import CausalImpact

DATA_DIR = os.path.join("rq1", "data")

MILESTONES = {
    "Oct2023": pd.Timestamp("2023-10-23"),
    "Mar2025": pd.Timestamp("2025-03-17"),
    "Jul2025": pd.Timestamp("2025-07-21"),
}

GROUPS = ["VPN", "Politics"]
FILTERS = ["Unmatched (non-UK)", "Unmatched", "Unmatched (UK)",
           "Raw", "Raw (UK)", "Matched", "Matched (UK)",
           "Classified", "Classified (UK)"]
METRICS = ["posts", "comments", "total"]


def run_ci(values, weeks, post_start):
    df = pd.DataFrame({"y": values}, index=weeks).astype(float)
    pre_end = post_start - pd.Timedelta(days=7)
    post_end = post_start + pd.Timedelta(days=7 * 7)
    ci = CausalImpact(df, [df.index.min(), pre_end], [post_start, post_end], alpha=0.05)
    s = ci.summary()
    rel = p = None
    for line in s.split("\n"):
        if "Relative effect" in line and rel is None:
            for tok in line.split():
                try: rel = float(tok.strip("%()").replace(",", "")); break
                except: pass
        if "tail-area" in line:
            for tok in line.split():
                try:
                    v = float(tok)
                    if 0 <= v <= 1: p = v
                except: pass
    return rel, p


if __name__ == "__main__":
    weekly = pd.read_csv(os.path.join(DATA_DIR, "table13_weekly_series.csv"))
    weekly["week"] = pd.to_datetime(weekly["week"])

    total = len(GROUPS) * len(FILTERS) * len(MILESTONES) * len(METRICS)
    print(f"Running {total} Table 13 models...\n")

    rows = []
    count = 0
    for group in GROUPS:
        for filt in FILTERS:
            sub = weekly[(weekly["group"] == group) & (weekly["filter"] == filt)].sort_values("week")
            if len(sub) == 0: continue
            for ms_name, ms_date in MILESTONES.items():
                for metric in METRICS:
                    count += 1
                    print(f"[{count:3d}/{total}] {group:8s} {filt:22s} {metric:8s} {ms_name}", end=" ", flush=True)
                    try:
                        rel, p = run_ci(sub[metric], sub["week"], ms_date)
                        sig = "**" if p and p < 0.05 else ("*" if p and p < 0.10 else "")
                        print(f"{rel:+.0f}%{sig}" if rel else "ERROR")
                    except Exception as e:
                        rel, p = None, None
                        print(f"ERROR")
                    rows.append({"group": group, "filter": filt, "metric": metric,
                                 "milestone": ms_name,
                                 "relative_effect_pct": round(rel, 1) if rel else None,
                                 "p_value": round(p, 4) if p else None})

    out = pd.DataFrame(rows)
    out.to_csv(os.path.join(DATA_DIR, "table13_reproduced.csv"), index=False)
    print(f"\nSaved: {DATA_DIR}/table13_reproduced.csv ({len(out)} rows)")